cmd/clusterctl/pkg/client/cluster/move.go

```diff
@@ -202,11 +232,412 @@ func (o *objectGraph) Discovery(namespace string, types []metav1.TypeMeta) error
+
+	cFrom, err := o.fromProxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	// Get the source object
+	sourceObj := &unstructured.Unstructured{}
+	sourceObj.SetAPIVersion(nodeToCreate.identity.APIVersion)
+	sourceObj.SetKind(nodeToCreate.identity.Kind)
+	targetObjKey := client.ObjectKey{
+		Namespace: nodeToCreate.identity.Namespace,
+		Name:      nodeToCreate.identity.Name,
+	}
+
+	if err := cFrom.Get(ctx, targetObjKey, sourceObj); err != nil {
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			sourceObj.GroupVersionKind(), sourceObj.GetNamespace(), sourceObj.GetName())
+	}
+
+	targetObj := sourceObj.DeepCopy()
+
+	// New objects cannot have a specified resource version. Clear it out.
+	targetObj.SetResourceVersion("")
+
+	// Removes current OwnerReferences
+	targetObj.SetOwnerReferences(nil)
+
+	// Creates a new OwnerReferences towards the OwnerNode, if any.
+	if ownerNode != nil {
+		ownerRef := metav1.OwnerReference{
+			APIVersion: ownerNode.identity.APIVersion,
+			Kind:       ownerNode.identity.Kind,
+			Name:       ownerNode.identity.Name,
+			UID:        ownerNode.newUID, // Use the owner's newUID read from the target management cluster (instead of the UID read during discovery).
+		}
+
+		// Restores the attributes of the OwnerReference.
+		if attributes, ok := ownerNode.dependents[nodeToCreate]; ok {
+			ownerRef.Controller = attributes.Controller
+			ownerRef.BlockOwnerDeletion = attributes.BlockOwnerDeletion
+		}
+
+		targetObj.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
+	}
+
+	// If the nodeToCreate is a cluster, reset Cluster.Spec.Paused it will be properly reconciled in the target management cluster
+	if targetObj.GetAPIVersion() == clusterv1.GroupVersion.String() && targetObj.GetKind() == "Cluster" {
+		obj := targetObj.Object
+		if err := unstructured.SetNestedField(obj, false, "spec", "paused"); err != nil {
+			return errors.Wrapf(err, "failed to reset Cluster.Spec.Paused for  %q %s/%s",
+				targetObj.GroupVersionKind(), targetObj.GetNamespace(), targetObj.GetName())
+		}
+		targetObj.SetUnstructuredContent(obj)
+	}
+
+	// Creates the targetObj into the target management cluster.
+	cTo, err := o.toProxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	if err := cTo.Create(ctx, targetObj); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "error creating %q %s/%s",
+				targetObj.GroupVersionKind(), targetObj.GetNamespace(), targetObj.GetName())
+		}
+
+		// If the object already exists, try to update it.
+		// Nb. This should not happen, but it is supported to make move more resilient to unexpected interrupt/restarts of the move process.
+		o.log.V(2).Info("Object already exists, updating", nodeToCreate.identity.Kind, nodeToCreate.identity.Name, "Namespace", nodeToCreate.identity.Namespace)
+
+		// Retries the UID and the resource version for the update.
+		existingTargetObj := targetObj.DeepCopy()
```

Comment:
It would probably be more efficient to create a new Unstructured & set the APIVersion & Kind?

Reply:
done